src/Microsoft.AspNetCore.Routing/Matchers/MatcherEndpoint.cs

```diff
@@ -36,14 +38,42 @@ public sealed class MatcherEndpoint : Endpoint
 
             Invoker = invoker;
             Template = template;
-            Values = new RouteValueDictionary(values);
+            ParsedTemlate = TemplateParser.Parse(template);
+            var mergedDefaults = GetDefaults(ParsedTemlate, new RouteValueDictionary(values));
+            Values = mergedDefaults;
             Order = order;
         }
 
         public int Order { get; }
         public Func<RequestDelegate, RequestDelegate> Invoker { get; }
         public string Template { get; }
-
         public IReadOnlyDictionary<string, object> Values { get; }
+
+        // Todo: needs review
+        public RouteTemplate ParsedTemlate { get; }
```

Comment:
I don't like the idea of putting `RouteTemplate` on any new public APIs because the design of it is kinda wack.

I wrote a **clean** version of `RouteTemplate` and the parsing code that we can use if necessary, but that kinda duplicates the concept so I'm not in love with that idea either.

Could we find a way to make this usage less prominent? (not a public property on a public class).

I'm also OK with doing this temporarily if there's an issue to clean it up.

Reply:
Filed https://github.com/aspnet/Routing/issues/566